### PR TITLE
Traffic snapshot

### LINKED ISSUE #10: Get historic info through the Telraam API
Rankings in streets (top 3 bike streets, ...) requires historic data. We need to check whether historic data can be obtained through the Telraam API (e.g. get data of the previous N months). If not available, it might be interesting to investigate setting up a database for our own website.

### PR DESCRIPTION
**Pull request by giokara**
Closes https://github.com/stijnvanhoey/telraam-bloemekeswijk/issues/10 and https://github.com/stijnvanhoey/telraam-bloemekeswijk/issues/11.

Fetches historic data in the background and computes a ranking of some metrics once ready. A user can visualize the metrics by switching on/off some buttons at the bottom of the map.

Some remaining working points that could be improved in subsequent PRs:

- [x] Speed limits are currently all set to 50. Needs to be revised.
~- [ ] The data is fetched every time the user switches from one page to another (e.g. from blog back to main). This should be avoided as the /traffic_snapshot calls are expensive.~ → To be done in a subsequent PR.
- [x] Switching from one metric to another, without first switching off the first metric, leads to unexpected behavior (no metric selected). I will look into resolving this ASAP.

220213-functional-metrics

### CODE REVIEW

**Comment by stijnvanhoey** [comment]
Really nice work!

> Speed limits are currently all set to 50. Needs to be revised.

Hardcode this? Where can we lookup these street limits (I can't find them on https://data.stad.gent). Otherwise I just cycle around to look for zone 30 (and maybe put these on the map).

> The data is fetched every time the user switches from one page to another (e.g. from blog back to main). This should be avoided as the /traffic_snapshot calls are expensive.

Probably using stores could solve this. I think we can merge and improve this in a follow up PR.

> Switching from one metric to another, without first switching off the first metric, leads to unexpected behavior (no metric selected). I will look into resolving this ASAP.

Any thoughts on how to fix this?

**Comment by giokara** [comment]
> Hardcode this?

I think that's the easiest, it can be done here: https://github.com/stijnvanhoey/telraam-bloemekeswijk/blob/7bce511a61694e20fddd0b904d3dab3f4a07832c/src/types.js#L147-L191. Also the last street added by Davy is part of this list.

> Otherwise I just cycle around to look for zone 30 (and maybe put these on the map).

Will probably be the easiest, happy to also do some streets. Google APIs contain that information, but they're paid (although very cheap) and I couldn't find any better way. The speed limit is now shown on the popup when no metric is chosen (might be interesting to add to the popup for speed violation metrics as well). Would be interesting to have this as an extra metric for visualization?

> Probably using stores could solve this. I think we can merge and improve this in a follow up PR.

:+1:

> Any thoughts on how to fix this?

Should be fixed now. The state management is quite hacky though. I'll try to improve this in the future.

**Comment by giokara** [comment]
> Where can we lookup these street limits

Got some help from someone at mobiliteit. The speed limits should be correct now.

**Comment by giokara-oqton** [comment]
@stijnvanhoey is there a possibility to increase the rate limits for the speed-... api call? I tried out the preview and seems that the rate limit is blocking fetching the traffic_snapshot data. Note that per user, now 11 speed-... calls will be made (1 / segment).

**Comment by giokara** [comment]
@stijnvanhoey is there a possibility to increase the rate limits for the speed-... api call? I tried out the preview and seems that the rate limit is blocking fetching the traffic_snapshot data. Note that per user, now 11 speed-... calls will be made (1 / segment).

**Comment by stijnvanhoey** [comment]
> @stijnvanhoey is there a possibility to increase the rate limits for the speed-... api call? I tried out the preview and seems that the rate limit is blocking fetching the traffic_snapshot data. Note that per user, now 11 speed-... calls will be made (1 / segment).

Yes I think we can up to the level we hit the limit of telraam API and we'll have to be 'smarter' on when to fetch which data/ I adjusted the environmental var to 50 in production. 

_on https://documenter.getpostman.com/view/8210376/TWDRqyaV:

> The following limits apply to the usage of the API:
> - Rate: 1 request per second
> - Burst: 1 request
> - Quota: 1000 requests per day

File: src/lib/components/MapPopup.svelte [hunk 1]
```diff
@@ -1,104 +1,156 @@
...
+	}
 
 	let options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
 	const currentDateStr = new Date(Date.parse(currentDateString)).toLocaleString('nl-BE', options);
 	const currentHourNum = new Date(Date.parse(currentDateString)).getHours();
-	const previousHourNum = currentHourNum - 1;
-	const isInactive = properties.state === SegmentStateEnum.INACTIVE;
-	const isNew = properties.state === SegmentStateEnum.NEW;
+	const previousDateStr = new Date(Date.parse(previousDateString)).toLocaleString('nl-BE', options);
+	const previousHourNum = new Date(Date.parse(previousDateStr)).getHours();
```
Comment (stijnvanhoey) [inline src/lib/components/MapPopup.svelte]: minor bug to fix the hour parsing: `previousDateStr`  -> `previousDateString`

File: src/routes/index.svelte [hunk 2]
```diff
@@ -55,28 +47,127 @@
...
-	//         console.log(speed.report);
-	//         speedData = speed.report;
-	//     }).catch(error => {
-	//         console.log(error);
-	//     })
-	// );
-	//$: segments = snapshot.features.map(x => x.properties.segment_id)
+	const timeEnd = new Date();
+	let timeStart = new Date(timeEnd);
+	timeStart.setMonth(timeStart.getMonth() - 1);
```
Comment (stijnvanhoey) [inline src/routes/index.svelte]: Is the monthly counts the one we are most interested in? Or would 'last week' make more sense?

**Comment by stijnvanhoey** [comment]
@giokara I like the adjustments. Having the integration of speed limits is a really good idea. 

Although, two thinks I can't really grasp:

- On what basis/values do you set the color scale?
- I can't find difference between the <20% and > 20% above speed limit effective values? Should there be a difference? I do understand the option for two bars, but I'm not sure the people will grasp it easily. 

I propose to convert the options to just 2: 

- 1/ zwakke/sterke weggebruikers  -> link to diverging color scale which suggests intensity in two directions instead of green/red
image

- 2/ speeding in which we integrate the <20% and > 20% speeding. It is a 2-D issue to represent: on one end % above speed limit and the other end % of drivers speeding above that limit. The 'data-viz' way of handling this a bivariate colormap, e.g. https://observablehq.com/@d3/bivariate-choropleth but this is also not supereasy to understand. We maybe can split it into categories together and maybe work with a color palette like this: image
e.g. light yellow -> 0% drivers exceed speed limit; orange -> 10% drivers exceed with speed limit with < 20%; dark brown > 20% drivers exceed speed limit with > 20%.

=> we'll have to add a legend on the map, but that should be fine with leaflet

What do you think?

File: src/lib/components/MapPopup.svelte (hunk 1 above)
Comment (giokara-oqton) [reply src/lib/components/MapPopup.svelte]: Thanks for noticing this!

Comment (giokara-oqton) [reply src/lib/components/MapPopup.svelte]: Thanks for noticing this!

File: src/routes/index.svelte (hunk 2 above)
Comment (giokara-oqton) [reply src/routes/index.svelte]: Updated, it will also make the website a bit more responsive.

File: src/lib/components/MapPopup.svelte (hunk 1 above)
Comment (giokara) [reply src/lib/components/MapPopup.svelte]: Thanks for noticing this!

File: src/routes/index.svelte (hunk 2 above)
Comment (giokara) [reply src/routes/index.svelte]: Updated, it will also make the website a bit more responsive.

**Comment by giokara** [comment]
> On what basis/values do you set the color scale?

This was based on the rank (1st = green, 2nd = orange, 3+ = red). But based on your comments, I gave a try to introduce a colormap.

> 2/ speeding in which we integrate the <20% and > 20% speeding.

I still kept two separate metrics, mainly due to laziness... But hopefully they're a bit more intuitive: "% chauffeurs binnen snelheidslimiet" (positive connotation 😄 ) and "Gemiddelde snelheid bij snelheidsovertreding". I also switched to a colorbar with 6 colors (red, red-orange, orange, yellow-orange, yellow, yellow-green).
I do like the idea of more intricate colorschemes, since the "Gemiddelde snelheid bij snelheidsovertreding" (expressed in "km/u boven snelheidslimiet") provides negative insights that might reflect very badly for the street situation, while being caused by a small minority of violators... Any further contributions to this, either in this PR or in a subsequent one, is very welcome!

Also a colorbar is still missing. Any idea how to place this on top of the leaflet map?

**Comment by giokara** [comment]
> Rate: 1 request per second

Hmm, does that mean we need to do the traffic_snapshot requests 1 by 1 with a delay of 1 s in between? I implemented this now this way by introduction of a function called `chainFetches`, but this does not seem very scalable... With the current amount of streets, waiting time will be a minimum of 12 seconds, excluding the actual API response times.

Is it possible to deploy the latest version of this PR as you did before [here](https://github.com/stijnvanhoey/telraam-bloemekeswijk/pull/16#issuecomment-1038798586) in order to test the rate-limiting and performance?

**Comment by stijnvanhoey** [comment]
> Hmm, does that mean we need to do the traffic_snapshot requests 1 by 1 with a delay of 1 s in between?

I don't think so, as we did more request already before. I would  just try and see how far we get.

> Is it possible to deploy the latest version of this PR 

done

**Comment by stijnvanhoey** [comment]
> I still kept two separate metrics, mainly due to laziness... But hopefully they're a bit more intuitive: "% chauffeurs binnen snelheidslimiet" (positive connotation smile ) and "Gemiddelde snelheid bij snelheidsovertreding"

I like!

**Comment by stijnvanhoey** [comment]
> Also a colorbar is still missing. Any idea how to place this on top of the leaflet map?

no :-(

**Comment by giokara-oqton** [comment]
> I don't think so, as we did more request already before. I would just try and see how far we get.

I did some further testing by providing a function `chainFetches` that can fetch in chunks. Here are the results

| Chunk size  | Total time for all /traffic_snaphot requests (ms) | 429 occurs |
| ------------- | ------------- | ---------- |
|  Unlimited (16 segments)  | 3119  | Yes |
| 10  | 4490  | Yes |
| 5 | 7449 | No |
| 3 | 8070 | No |
| 1 | 31371 | No |

Current implementation therefore does 3 requests in parallel in order to avoid 429 errors, while still being snappier than chaining single calls.

**Comment by giokara** [comment]
> I don't think so, as we did more request already before. I would just try and see how far we get.

I did some further testing by providing a function `chainFetches` that can fetch in chunks. Here are the results

| Chunk size  | Total time for all /traffic_snaphot requests (ms) | 429 errors |
| ------------- | ------------- | ---------- |
|  Unlimited (16 segments)  | 3119  | Yes |
| 10  | 4490  | Yes |
| 5 | 7449 | No |
| 3 | 8070 | No |
| 1 | 31371 | No |

Current implementation therefore does 3 requests in parallel in order to avoid 429 errors, while still being snappier than chaining single calls.

**Comment by giokara** [comment]
> Also a colorbar is still missing. 

Added a colorbar by using some MIT licensed implementation.
<p float="left">
Screenshot 2022-03-19 at 12 19 31
Screenshot 2022-03-19 at 12 45 31
</p>

**Comment by giokara** [comment]
> Is it possible to deploy the latest version of this PR

For some reason I'm not able to see the added controls below the map in the deployed version. Are you seeing them @stijnvanhoey? Also no errors in the console, so no idea why this is not working.

Not planning any further changes to this PR from my side. If OK, we can try to deploy this.

**Comment by stijnvanhoey** [comment]
> For some reason I'm not able to see the added controls below the map in the deployed version.

Indeed. 

Certainly fine for me to deploy; nice work!

**Pull request merged**